File: api/src/controllers/health.ts
import { Request, Response } from 'express';
import { dataSource } from 'database/createConnection';
import { isJiraIntegrationsConfigured } from 'config';
import { logger } from 'utils/logger';

interface HealthCheckResult {
  name: string;
  status: 'healthy' | 'unhealthy';
  details?: string;
}

export async function healthCheck(_req: Request, res: Response): Promise<void> {
  const checks: HealthCheckResult[] = [];
  let isHealthy = true;

  const dbCheck: HealthCheckResult = {
    name: 'database',
    status: 'healthy',
  };

  try {
    if (dataSource.isInitialized) {
      await dataSource.query('SELECT 1');
    } else {
      dbCheck.status = 'unhealthy';
      dbCheck.details = 'Database connection not initialized';
      isHealthy = false;
    }
  } catch (error) {
    dbCheck.status = 'unhealthy';
    dbCheck.details = error instanceof Error ? error.message : 'Unknown database error';
    isHealthy = false;
    logger.error({ error: error instanceof Error ? error.message : error }, 'Health check: database error');
  }

  checks.push(dbCheck);

  const jiraCheck: HealthCheckResult = {
    name: 'jira-integrations',
    status: isJiraIntegrationsConfigured() ? 'healthy' : 'unhealthy',
    details: isJiraIntegrationsConfigured() 
      ? undefined 
      : 'Jira integration not configured',
  };
  checks.push(jiraCheck);

  const response = {
    status: isHealthy ? 'healthy' : 'unhealthy',
    timestamp: new Date().toISOString(),
    checks,
  };

  res.status(isHealthy ? 200 : 503).json(response);
}

export async function readinessCheck(_req: Request, res: Response): Promise<void> {
  try {
    if (!dataSource.isInitialized) {
      res.status(503).json({ status: 'not_ready', reason: 'Database not initialized' });
      return;
    }

    await dataSource.query('SELECT 1');
    
    res.status(200).json({ status: 'ready' });
  } catch (error) {
    logger.error({ error: error instanceof Error ? error.message : error }, 'Readiness check failed');
    res.status(503).json({ 
      status: 'not_ready', 
      reason: error instanceof Error ? error.message : 'Unknown error' 
    });
  }
}

export async function livenessCheck(_req: Request, res: Response): Promise<void> {
  res.status(200).json({ status: 'alive', timestamp: new Date().toISOString() });
}

File: packages/planny-api/src/planny_api/routers/health.py
"""Health check router — liveness, readiness, and full health probes.

Matches the Node ``/health``, ``/health/ready``, ``/health/live`` endpoints.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

import structlog
from fastapi import APIRouter, Depends
from planny_core.config import settings
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from planny_api.dependencies import get_db

logger = structlog.get_logger()
router = APIRouter(tags=["health"])


async def _check_database(db: AsyncSession) -> dict[str, Any]:
    """Check database connectivity with a simple ``SELECT 1``."""
    try:
        await db.execute(text("SELECT 1"))
        return {"name": "database", "status": "ok"}
    except Exception as exc:
        logger.warning("Health check: database unreachable", error=str(exc))
        return {"name": "database", "status": "error", "error": "Database connection failed"}


def _check_jira_integrations() -> dict[str, Any]:
    """Check whether Jira integrations are configured (non-empty env)."""
    internal_ok = bool(settings.internal_atlassian_base_url)
    external_ok = bool(settings.external_atlassian_base_url)
    status = "ok" if (internal_ok or external_ok) else "not_configured"
    return {
        "name": "jira_integrations",
        "status": status,
        "internal_configured": internal_ok,
        "external_configured": external_ok,
    }


@router.get("/health")
async def health_check(db: AsyncSession = Depends(get_db)) -> dict[str, Any]:
    """Full health check — database + Jira integration status."""
    db_check = await _check_database(db)
    jira_check = _check_jira_integrations()
    overall = "healthy" if db_check["status"] == "ok" else "unhealthy"

    return {
        "status": overall,
        "timestamp": datetime.now(UTC).isoformat(),
        "checks": [db_check, jira_check],
    }


@router.get("/health/ready")
async def readiness_check(db: AsyncSession = Depends(get_db)) -> dict[str, Any]:
    """Readiness probe — database connectivity only."""
    db_check = await _check_database(db)
    status = "ready" if db_check["status"] == "ok" else "not_ready"

    return {
        "status": status,
        "timestamp": datetime.now(UTC).isoformat(),
    }


@router.get("/health/live")
async def liveness_check() -> dict[str, Any]:
    """Liveness probe — always returns OK."""
    return {
        "status": "alive",
        "timestamp": datetime.now(UTC).isoformat(),
    }
